### PR TITLE
support "grpc-status-details-bin"

### LINKED ISSUE #204: transcoding: support "grpc-status-details-bin"
Seems grpc-go sends "grpc-status-details-bin" in trailers to have error details (https://github.com/grpc/grpc-go/blob/v1.4.1/transport/http2_server.go#L750). We might need to support that in transcoding.

TODO:
- Confirm what is the standard format for "grpc-status-details-bin", it is not documented in [gRPC Wire Protocol](http://www.grpc.io/docs/guides/wire.html).
- Confirm C and Java stack behavior.

### PR DESCRIPTION
**Pull request by JimmyCYJ**
Which issue this PR fixes: fixes #204

### CODE REVIEW

File: include/api_manager/utils/status.h [hunk 1]
```diff
@@ -115,6 +120,11 @@ class Status final {
 
   // Error cause indicating the origin of the error.
   ErrorCause error_cause_;
+
+  // Detailed information of a grpc response error. If gRPC response trailers 
+  // have a "grpc-status-details-bin" header, the decoded value of that 
+  // header is stored here.
+  std::string grpc_status_details_;
```
Comment (qiwzhang) [inline include/api_manager/utils/status.h]: I don't think this is correct.  My understanding:  "grpc-status-detail-bin" is a proto binary of rpc:status which is error_code and error_message.  We need to:
* get the header value, base64 decode it
* call ParseFromString() to convert binary into proto
* assign the code and error_message

File: src/nginx/error.cc [hunk 2]
```diff
@@ -207,8 +208,19 @@ ngx_int_t ngx_esp_error_body_filter(ngx_http_request_t *r, ngx_chain_t *in) {
         // Serialize error as JSON
         ngx_buf_t *body = nullptr;
         ngx_str_t json_error;
-        if (ngx_str_copy_from_std(r->pool, ctx->status.ToJson(), &json_error) !=
-            NGX_OK) {
+
+        if (ctx->grpc_status_details != nullptr) {
+          std::string status_details_in_json;
+          utils::Status::StatusProtoToJson(
+              *(ctx->grpc_status_details), &status_details_in_json,
```
Comment (qiwzhang) [inline src/nginx/error.cc]: remove "("   bracket

File: src/nginx/module.h [hunk 3]
```diff
@@ -248,6 +248,11 @@ struct ngx_esp_request_ctx_s {
 
   // Nginx variable $backend_url. set from service config.
   ngx_str_t backend_url;
+
+  // If gRPC response have a "grpc-status-details-bin" header, which
```
Comment (qiwzhang) [inline src/nginx/module.h]: Update the comment as
// Stores proto message of gRPC response header "grpc-status-details-bin"

File: src/nginx/error.cc [hunk 4]
```diff
@@ -207,8 +208,19 @@ ngx_int_t ngx_esp_error_body_filter(ngx_http_request_t *r, ngx_chain_t *in) {
         // Serialize error as JSON
         ngx_buf_t *body = nullptr;
         ngx_str_t json_error;
-        if (ngx_str_copy_from_std(r->pool, ctx->status.ToJson(), &json_error) !=
-            NGX_OK) {
+
+        if (ctx->grpc_status_details != nullptr) {
```
Comment (qiwzhang) [inline src/nginx/error.cc]: Add comment
// if there is grpc-status-detail-bin response header, generate error json body with that header.

File: src/nginx/t/transcoding_errors.t [hunk 5]
```diff
@@ -137,7 +137,7 @@ $response = ApiManager::http_get($NginxPort,'/shelves/100?key=api-key-3');
 
 like($response, qr/HTTP\/1\.1 404 Not Found/, 'Got a 404.');
 like($response, qr/Content-Type: application\/json/i, 'Content-type is application/json');
-like($response, qr/Shelf not found/i, "the message from the backend was propogated.");
+like($response, qr/Cannot find shelf 100/i, "the message from the backend was propogated.");
```
Comment (qiwzhang) [inline src/nginx/t/transcoding_errors.t]: add comment that this error message is from detail_status from gRPC backend.

File: src/api_manager/utils/status.cc [hunk 6]
```diff
@@ -239,6 +239,24 @@ bool Status::operator==(const Status& x) const {
                 proto_status.error_message().ToString());
 }
 
+/* static */ void Status::StatusProtoToJson(
```
[the remark this answers is not in the document]
Comment (JimmyCYJ) [reply src/api_manager/utils/status.cc]: Removed. Thanks.

File: src/nginx/error.cc (hunk 4 above)
Comment (JimmyCYJ) [reply src/nginx/error.cc]: Added.

File: src/nginx/module.h (hunk 3 above)
Comment (JimmyCYJ) [reply src/nginx/module.h]: Done.

File: src/nginx/t/transcoding_errors.t (hunk 5 above)
Comment (JimmyCYJ) [reply src/nginx/t/transcoding_errors.t]: Done.

**Comment by JimmyCYJ** [comment]
Please take a look. Thanks.

**Pull request merged**